instrumentation/opentelemetry_oban/lib/opentelemetry_oban/job_handler.ex

```diff
@@ -0,0 +1,101 @@
+defmodule OpentelemetryOban.JobHandler do
+  alias OpenTelemetry.Span
+
+  @tracer_id :opentelemetry_oban
+
+  def attach() do
+    attach_job_start_handler()
+    attach_job_stop_handler()
+    attach_job_exception_handler()
+  end
+
+  defp attach_job_start_handler() do
+    :telemetry.attach(
+      "#{__MODULE__}.job_start",
+      [:oban, :job, :start],
+      &__MODULE__.handle_job_start/4,
+      []
+    )
+  end
+
+  defp attach_job_stop_handler() do
+    :telemetry.attach(
+      "#{__MODULE__}.job_stop",
+      [:oban, :job, :stop],
+      &__MODULE__.handle_job_stop/4,
+      []
+    )
+  end
+
+  defp attach_job_exception_handler() do
+    :telemetry.attach(
+      "#{__MODULE__}.job_exception",
+      [:oban, :job, :exception],
+      &__MODULE__.handle_job_exception/4,
+      []
+    )
+  end
+
+  def handle_job_start(_event, _measurements, metadata, _config) do
+    %{
+      job: %{
+        id: id,
+        queue: queue,
+        worker: worker,
+        priority: priority,
+        inserted_at: inserted_at,
+        scheduled_at: scheduled_at,
+        attempt: attempt,
+        max_attempts: max_attempts,
+        meta: job_meta
+      }
+    } = metadata
+
+    :otel_propagator_text_map.extract(Map.to_list(job_meta))
+    parent = OpenTelemetry.Tracer.current_span_ctx()
+    links = if parent == :undefined, do: [], else: [OpenTelemetry.link(parent)]
+    OpenTelemetry.Tracer.set_current_span(:undefined)
```

Comment:
@indrekj Due to the force pushes on this PR I lost a bit track when this was introduced (maybe it's a good idea to do one force push at the end ;)), but currently I'm having the problem that the span started inside a job are not children of the job span, is this intended? Are you also seeing this behaviour? The consequence of this is that my custom sampler is not that useful anymore :)

Reply:
I mean, I can always start another wrapper span inside the job, but I just wanted to double check and understand that this is intended behavior.

Reply:
We're not seeing this. I also added a test to ensure it: https://github.com/open-telemetry/opentelemetry-erlang-contrib/compare/db0e1f1a67bbd51992b10a0020da3b21119f6edb..ddf26187cee991e329c9dcbeaa42ad50eacc0c5d

We however had a similar issue when we were using `def timeout(_job) do` in the jobs. If timeout function is defined then oban uses perform_timed to execute the worker, see: https://github.com/sorentwo/oban/blob/60274b76ff2c59d660ae1bf7721a3ff14854700b/lib/oban/queue/executor.ex#L257-L269 . This creates a new process so the current span information is lost. We only had one job in the entire system with that method, so we used:

```
  # We use our own `with_timeout` instead of Oban.Worker.timeout/1 because if
  # the latter is used then the trace context is not propagated.
  defp with_timeout(fun) do
    task = Task.async(OpentelemetryFunction.wrap(fun))

    case Task.yield(task, @timeout) || Task.shutdown(task) do
      {:ok, reply} ->
        reply

      nil ->
        {:error, :timeout}
    end
  end
```

Reply:
I'd like to get this PR merged to start addressing these kind of rare edge cases. As you said, the PR is huge and it's hard to follow the changes now.

Reply:
Ahhh, yes in this case we also used `Oban.Worker.timeout/1`, thanks! 👍 

> I'd like to get this PR merged to start addressing these kind of rare edge cases. As you said, the PR is huge and it's hard to follow the changes now.

I agree, maybe we can track these open issues / edge cases in the description so we can work on it when the PR is merged.

Reply:
Yea, I can merge and then you should start opening issues.

I saw some question about `set_attributes` and sampling but can't find it again, so answering here: The sampler won't see attributes set with `set_attributes` because it has already been sampled at that point. A span is started and the attributes passed to the function that starts the span will be available but any attributes set after starting the span are of course not available.

Not sure what the code was, maybe it was something that used a pipe to look like it was using the "builder pattern" to construct a span? But we don't use that pattern, so there is no "build" function that would come after doing something like `span("name") |> set_attributes(...) |> start()`. So yea, must pass any attributes that need to be available for sampling to the start call.

Reply:
Thanks. It was here: https://github.com/open-telemetry/opentelemetry-erlang-contrib/pull/6/files/9ff9f64044d47306c0764dd9e12472a1b7eb3349#diff-ca8cd6f4f23c284564be4b34cb5caf868177e4190300720bc3c6fdb255c18b90 . For that particular case we were able to avoid set_attributes

> Yea, I can merge and then you should start opening issues.

:+1: